base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs

```diff
@@ -0,0 +1,886 @@
+            DbKeyValuePair,
+            DbValue,
+            OutputManagerBackend,
+            PendingTransactionOutputs,
+            WriteOperation,
+        },
+        TxId,
+    },
+    schema::{outputs, pending_transaction_outputs},
+};
+use chrono::{Duration as ChronoDuration, NaiveDateTime, Utc};
+use diesel::{
+    prelude::*,
+    r2d2::{ConnectionManager, Pool, PooledConnection},
+    result::Error as DieselError,
+    SqliteConnection,
+};
+use std::{collections::HashMap, convert::TryFrom, io, path::Path, time::Duration};
+use tari_transactions::{
+    tari_amount::MicroTari,
+    transaction::{OutputFeatures, OutputFlags, UnblindedOutput},
+    types::PrivateKey,
+};
+use tari_utilities::ByteArray;
+
+const DATABASE_CONNECTION_TIMEOUT_MS: u64 = 2000;
+
+/// A Sqlite backend for the Output Manager Service. The Backend is accessed via a connection pool to the Sqlite file.
+pub struct OutputManagerSqliteDatabase {
+    database_connection_pool: Pool<ConnectionManager<SqliteConnection>>,
+}
+impl OutputManagerSqliteDatabase {
+    pub fn new(database_path: String) -> Result<Self, OutputManagerStorageError> {
+        let db_exists = Path::new(&database_path).exists();
+
+        let connection = SqliteConnection::establish(&database_path)?;
+
+        connection.execute("PRAGMA foreign_keys = ON")?;
+        if !db_exists {
+            embed_migrations!("./migrations");
+            embedded_migrations::run_with_output(&connection, &mut io::stdout()).map_err(|err| {
+                OutputManagerStorageError::DatabaseMigrationError(format!("Database migration failed {}", err))
+            })?;
+        }
+        drop(connection);
+
+        let manager = ConnectionManager::<SqliteConnection>::new(database_path);
+        let pool = diesel::r2d2::Pool::builder()
+            .connection_timeout(Duration::from_millis(DATABASE_CONNECTION_TIMEOUT_MS))
+            .idle_timeout(Some(Duration::from_millis(DATABASE_CONNECTION_TIMEOUT_MS)))
+            .build(manager)
+            .map_err(|_| OutputManagerStorageError::R2d2Error)?;
+
+        Ok(Self {
+            database_connection_pool: pool,
+        })
+    }
+}
+impl OutputManagerBackend for OutputManagerSqliteDatabase {
+    fn fetch(&self, key: &DbKey) -> Result<Option<DbValue>, OutputManagerStorageError> {
+        let conn = self
+            .database_connection_pool
+            .clone()
+            .get()
+            .map_err(|_| OutputManagerStorageError::R2d2Error)?;
+
+        let result = match key {
+            DbKey::SpentOutput(k) => match OutputSql::find_spent(&k.to_vec(), true, &conn) {
+                Ok(o) => Some(DbValue::SpentOutput(Box::new(UnblindedOutput::try_from(o)?))),
+                Err(e) => {
+                    match e {
+                        OutputManagerStorageError::DieselError(DieselError::NotFound) => (),
+                        e => return Err(e),
+                    };
+                    None
+                },
```

Comment:
I think this code can be simplified by removing the second match, by doing something like this:
` Ok(o) => Some(DbValue::SpentOutput(Box::new(UnblindedOutput::try_from(o)?))),`
` Err( OutputManagerStorageError::DieselError(DieselError::NotFound)) => None,`
` Err(e) => Err(e),`
I didn't test this code, but something like this.

Reply:
The issue is that all match branch need to return the same type so you cannot have one return None and another Err(e) and you cannot have the other branches return options while that one branch has a `return Err(e)` because that is not an option. There might be a sneaky way to flatten this but I couldn't see it.